### PR TITLE
kie-issues#947: Playwright expansion phase - part 2 - Decision Service node

### PR DESCRIPTION
**Pull request by jomarko**
This is another part of the the playwright testsuite expansion phase https://github.com/apache/incubator-kie-tools/issues/3905

### CODE REVIEW

File: packages/dmn-editor/src/draggable/Draggable.tsx [hunk 1]
```diff
@@ -252,24 +252,27 @@ export function Draggable(props: {
...
-          onPointerLeave={() => setDraggable(false)}
-          style={props.handlerStyle}
-        >
-          <GripVerticalIcon
-            className={
-              hovered ? "kie-dmn-editor--draggable-icon-handler-hovered" : "kie-dmn-editor--draggable-icon-handler"
-            }
-          />
-        </Icon>
+        <div data-testid={"kie-dmn-editor--draggable-icon"}>
```
Comment (ljmotta) [inline packages/dmn-editor/src/draggable/Draggable.tsx]: We should use `kie-tools--dmn-editor--` prefix for `data-testid`. Please review all `data-testids`

File: packages/dmn-editor/src/propertiesPanel/DecisionServiceProperties.tsx [hunk 2]
```diff
@@ -163,13 +163,15 @@ export function DecisionServiceProperties({
 
       <FormGroup label="Output decisions">
         <DecisionServiceElementList
+          elementListId={"kie-tools--dmn-editor--decision-service-output-decisions"}
```
Comment (ljmotta) [inline packages/dmn-editor/src/propertiesPanel/DecisionServiceProperties.tsx]: Please, use `elementListDataTestId` to avoid ambiguity. The same applies for all occurences.

File: packages/dmn-editor/tests-e2e/__fixtures__/propertiesPanel/decisionServicePropertiesPanel.ts [hunk 3]
```diff
@@ -120,4 +120,71 @@ export class DecisionServicePropertiesPanel extends PropertiesPanelBase {
...
+      await this.panel()
+        .getByTestId("kie-tools--dmn-editor--decision-service-input-decisions")
+        .getByTestId("kie-dmn-editor--draggable-children")
+        .allTextContents()
+    ).map((content) => content.trim());
+  }
+
+  public async moveInputDecision(args: { nth: number; way: "up" | "down" }) {
+    // TODO
+  }
```
Comment (ljmotta) [inline packages/dmn-editor/tests-e2e/__fixtures__/propertiesPanel/decisionServicePropertiesPanel.ts]: It's a leftover? The `moveInputDecision` will be left for another PR?

File: packages/dmn-editor/tests-e2e/drgElements/modelDecisionService.spec.ts [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+      await diagram.resetFocus();
+    });
+
+    test("Decision Service Signature - Output Decisions", async ({
+      decisionServicePropertiesPanel,
+      diagram,
+      nodes,
+      palette,
+    }) => {
+      // https://github.com/apache/incubator-kie-issues/issues/663
```
Comment (ljmotta) [inline packages/dmn-editor/tests-e2e/drgElements/modelDecisionService.spec.ts]: This kind of information is useful on the test report, and we can add Annotations to the report with `test.info().annotations.push(...)`. The same applies to the other comments.

File: packages/dmn-editor/tests-e2e/drgElements/modelDecisionService.spec.ts [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+        // !!! CURRENTLY PUSHED ONE OF MANY ATTEMPTS
+        await decisionServicePropertiesPanel.moveInputData({ nth: 0, way: "down" });
+        await decisionServicePropertiesPanel.moveInputDecision({ nth: 0, way: "down" });
+        expect(await decisionServicePropertiesPanel.getInvokingThisDecisionServiceInFeel()).toEqual(
+          "New Decision Service(A, B, AA, BB)"
+        );
+        expect(await decisionServicePropertiesPanel.getInputDecisions()).toEqual(["A", "B"]);
+        expect(await decisionServicePropertiesPanel.getInputData()).toEqual(["AA", "BB"]);
+      });
+    });
```
Comment (ljmotta) [inline packages/dmn-editor/tests-e2e/drgElements/modelDecisionService.spec.ts]: What isn't specifically working? The re-order by clicking and draging?

File: packages/dmn-editor/tests-e2e/__fixtures__/propertiesPanel/decisionServicePropertiesPanel.ts [hunk 6]
```diff
@@ -120,4 +120,70 @@ export class DecisionServicePropertiesPanel extends PropertiesPanelBase {
...
+  public async getInputDecisions() {
+    return (
+      await this.panel()
+        .getByTestId("kie-tools--dmn-editor--decision-service-input-decisions")
+        .getByTestId("kie-dmn-editor--draggable-children")
+        .allTextContents()
+    ).map((content) => content.trim());
+  }
+
+  public async moveInputDecision(args: { nth: number; way: "up" | "down" }) {
```
Comment (ljmotta) [inline packages/dmn-editor/tests-e2e/__fixtures__/propertiesPanel/decisionServicePropertiesPanel.ts]: This type of reorder will enable to move only one up/down. Why not using `from/to`?
I mean, we can have a case like:
1. InputData-1
2. InputData-2
3. InputData-3
4. InputData-4

And reorder to:
1. InputData-1
2. InputData-4
3. InputData-2
4. InputData-3

**Review by ljmotta — CHANGES_REQUESTED** [review]
@jomarko Thanks for this PR. The tests are looking very good. I've made some comments inline.

File: packages/dmn-editor/src/draggable/Draggable.tsx (hunk 1 above)
Comment (jomarko) [reply packages/dmn-editor/src/draggable/Draggable.tsx]: done

File: packages/dmn-editor/src/propertiesPanel/DecisionServiceProperties.tsx (hunk 2 above)
Comment (jomarko) [reply packages/dmn-editor/src/propertiesPanel/DecisionServiceProperties.tsx]: sure, good suggestion, done

File: packages/dmn-editor/tests-e2e/__fixtures__/propertiesPanel/decisionServicePropertiesPanel.ts (hunk 3 above)
Comment (jomarko) [reply packages/dmn-editor/tests-e2e/__fixtures__/propertiesPanel/decisionServicePropertiesPanel.ts]: yes, leftover, implemented now

File: packages/dmn-editor/tests-e2e/drgElements/modelDecisionService.spec.ts (hunk 4 above)
Comment (jomarko) [reply packages/dmn-editor/tests-e2e/drgElements/modelDecisionService.spec.ts]: done 6f6aabd63f

File: packages/dmn-editor/tests-e2e/__fixtures__/propertiesPanel/decisionServicePropertiesPanel.ts (hunk 6 above)
Comment (jomarko) [reply packages/dmn-editor/tests-e2e/__fixtures__/propertiesPanel/decisionServicePropertiesPanel.ts]: good suggestion, done 317ef019cf

**Comment by jomarko** [comment]
Feedback incorporated

**Review by ljmotta — COMMENTED** [review]
@jomarko Thanks for addressing the comments. I've just one regarding the annotation option. If the test ensures the problem was solved, it should be a `REGRESSION`, if the test is affected by a bug (e.g. not working as expected) it should be the `AFFECTED_BY`, and if the test has a workaround because of another issue (it can be a bug too, but we add a workaround step) it should be `WORKAROUND_DUE_TO`. For this case, I guess it should be "REGRESSION"?

File: packages/dmn-editor/tests-e2e/drgElements/modelDecisionService.spec.ts [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+        await page.keyboard.press("ArrowDown");
+      }
+
+      await nodes.select({ name: DefaultNodeName.DECISION_SERVICE, position: NodePosition.TOP });
+      await decisionServicePropertiesPanel.open();
+      expect(await decisionServicePropertiesPanel.getOutputDecisions()).toEqual("New Decision");
+
+      await expect(diagram.get()).toHaveScreenshot("move-decision-in-decision-service-by-keyboard.png");
+    });
+  });
```
Comment (ljmotta) [inline packages/dmn-editor/tests-e2e/drgElements/modelDecisionService.spec.ts]: No problem with the test name choice here. But please, stick with one name pattern for the entire suite (decision-service). You can change the others if you think it's better, but remember that the name should be unique and the `describe` blocks should help us filter a type of test. Revisit the previous suites if needed.

Comment (jomarko) [reply packages/dmn-editor/tests-e2e/drgElements/modelDecisionService.spec.ts]: Thank you, I tried to incorporate the feedback. Please let me know @ljmotta .

**Review by ljmotta — APPROVED** [review]
@jomarko Thanks a lot for your patience! PR is looking great.

File: packages/dmn-editor/tests-e2e/drgElements/resizeDecisionService.spec.ts [hunk 8]
```diff
@@ -304,5 +304,28 @@ test.describe("Resize node - Decision Service", () => {
...
+      palette,
+    }) => {
+      test.info().annotations.push({
+        type: TestAnnotations.REGRESSION,
+        description: "https://github.com/apache/incubator-kie-issues/issues/881",
+      });
+      await diagram.resetFocus();
+      await palette.dragNewNode({ type: NodeType.DECISION, targetPosition: { x: 400, y: 100 } });
+      await diagram.resetFocus();
+      // move into Decision Service
```
Comment (klbowers) [inline packages/dmn-editor/tests-e2e/drgElements/resizeDecisionService.spec.ts:320]: Is this comment needed? I feel like most of the tests you've added in this PR move a node into a Decision Service

File: packages/dmn-editor/tests-e2e/drgElements/resizeDecisionService.spec.ts [hunk 9]
```diff
@@ -304,5 +304,28 @@ test.describe("Resize node - Decision Service", () => {
...
+      decisionServicePropertiesPanel,
+      diagram,
+      nodes,
+      palette,
+    }) => {
+      test.info().annotations.push({
+        type: TestAnnotations.REGRESSION,
+        description: "https://github.com/apache/incubator-kie-issues/issues/881",
+      });
+      await diagram.resetFocus();
```
Comment (klbowers) [inline packages/dmn-editor/tests-e2e/drgElements/resizeDecisionService.spec.ts]: Do we need to reset at the start of a test?

File: packages/dmn-editor/tests-e2e/drgElements/modelDecisionService.spec.ts [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+        nodes,
+        palette,
+      }) => {
+        test.skip(true, "https://github.com/apache/incubator-kie-issues/issues/896");
+        test.info().annotations.push({
+          type: TestAnnotations.AFFECTED_BY,
+          description: "https://github.com/apache/incubator-kie-issues/issues/896",
+        });
+
+        await diagram.resetFocus();
```
Comment (klbowers) [inline packages/dmn-editor/tests-e2e/drgElements/modelDecisionService.spec.ts]: Do we need to reset at the start of a test?

**Review by klbowers — APPROVED** [review]
Thanks for the PR Jozef, LGTM! 
Added a few nitpick comments inline, just wanted to clarify if there was a reason for resetting the focus at the start of the test, as mostly you seem to drag the new node and then reset focus.

File: packages/dmn-editor/tests-e2e/drgElements/resizeDecisionService.spec.ts (hunk 8 above)
Comment (jomarko) [reply packages/dmn-editor/tests-e2e/drgElements/resizeDecisionService.spec.ts:320]: I would like to keep this comment for now, currently we have API that works more with coordinates. I mean, after a while, reading the test line, `await nodes.move({ name: DefaultNodeName.DECISION, targetPosition: { x: 200, y: 200 } });` may be not so clear the coordinates are inside a decision service node.

If we, in future introduce playwright fixture method like: `nodes.moveInto({name, targetNodeName})` then for sure we can remove this comment. What do you think @kbowers-ibm ?

File: packages/dmn-editor/tests-e2e/drgElements/resizeDecisionService.spec.ts (hunk 9 above)
Comment (jomarko) [reply packages/dmn-editor/tests-e2e/drgElements/resizeDecisionService.spec.ts]: well spotted, removing

File: packages/dmn-editor/tests-e2e/drgElements/modelDecisionService.spec.ts (hunk 10 above)
Comment (jomarko) [reply packages/dmn-editor/tests-e2e/drgElements/modelDecisionService.spec.ts]: well spotted, removing

File: packages/dmn-editor/tests-e2e/drgElements/resizeDecisionService.spec.ts (hunk 8 above)
Comment (danielzhe) [reply packages/dmn-editor/tests-e2e/drgElements/resizeDecisionService.spec.ts:320]: I know you didn't ask for me, but since I'm here: I had the same "issue" with some tests I did. I think that `nodes.moveInto({name, targetNodeName})` would be a great idea but for now, since we have another test doing exactly the same thing, I think we should leave it as it is. Also, it can be very tricky to do a `nodes.moveInto({name, targetNodeName})`, because consider the scenarios when we have more than one node inside, different sizes, and so on. Probably the user of `nodes.moveInto({name, targetNodeName})` would be expected that the node is moved in a good position with no overlap.

Comment (klbowers) [reply packages/dmn-editor/tests-e2e/drgElements/resizeDecisionService.spec.ts:320]: Yeah sure makes sense, twas just a minor thing!

File: packages/dmn-editor/src/propertiesPanel/DecisionServiceProperties.tsx [hunk 11]
```diff
@@ -193,6 +196,7 @@ export function DecisionServiceProperties({
       </FormGroup>
       <FormGroup label="Input data">
         <DraggableDecisionServiceElementList
+          elementListDataTestId={"kie-tools--dmn-editor--decision-service-input-data"}
```
Comment (danielzhe) [inline packages/dmn-editor/src/propertiesPanel/DecisionServiceProperties.tsx]: I don't think it is a good practice to add the data-testid into the logic of the component by adding a new parameter, the `elementListDataTestId`.

What you can do is to add `data-testid` directly in the `FormGroup`:

`<FormGroup label="Input data" data-testid={"kie-tools--dmn-editor--decision-service-input-data"}>`

And then remove the `elementListDataTestId` from the `DraggableDecisionServiceElementList` and its nested `<ul>`.

If you need the list items inside the `"kie-tools--dmn-editor--decision-service-input-data"`, you can do something like this at the end:
`getByRole("listitem")`

Also, please, check the `DecisionServiceElementList` for the same thing.

**Review by danielzhe — CHANGES_REQUESTED** [review]

File: packages/dmn-editor/src/propertiesPanel/DecisionServiceProperties.tsx (hunk 11 above)
Comment (jomarko) [reply packages/dmn-editor/src/propertiesPanel/DecisionServiceProperties.tsx]: good suggestion @danielzhe , I like it, will incorporate it

Comment (jomarko) [reply packages/dmn-editor/src/propertiesPanel/DecisionServiceProperties.tsx]: @danielzhe changes pushed d9500cfff1

Comment (tiagobento) [reply packages/dmn-editor/src/propertiesPanel/DecisionServiceProperties.tsx]: Great catch @danielzhe and thanks for applying the suggestion @jomarko. Way to go!

**Review by danielzhe — APPROVED** [review]
Thank you! Great improvement to our test coverage!
